### PR TITLE
Wip/specialised conversion

### PR DESCRIPTION
**Pull request by kosmikus**
This is a PR as a basis for discussion.

The main motivation for this PR are `productFrom` and `productTo`, which I need often and find somewhat annoying and overly cryptic to write by hand. Since we have `IsProductType` now, I think it makes sense to also have dedicated conversion functions.

However, some things to consider:

- What about conversions for the other special cases? I've added them here, but I don't think I've ever needed them.

- What about names? In particular `enumFrom` clashes with a `Prelude` function. However, `fromEnum` isn't any better (and I think that name is misleading, too).

Feedback welcome.

### CODE REVIEW

**Comment by edsko** [comment]
Makes sense, seems consistent to leave them in. `sumFrom` might be an alternative to avoid the name clash and is consistent with `productFrom`.

**Comment by kosmikus** [comment]
@edsko Thanks for the suggestion. I decided against `IsSumType` and in favour of `IsProductType` back when I introduced the type synonyms because I'd usually call any type with multiple constructors a sum (of products) type, whereas here I want to restrict to sum types with only nullary constructors.

File: src/Generics/SOP/Universe.hs [hunk 1]
```diff
@@ -173,6 +191,20 @@ type IsProductType (a :: *) (xs :: [*]) =
 type IsEnumType (a :: *) =
   (Generic a, All ((~) '[]) (Code a))
 
+-- | Convert from an enum type to its sum representation.
+--
+-- @since 0.3.3.0
+--
+enumFrom :: IsEnumType a => a -> NS (K ()) (Code a)
+enumFrom = map_NS (const (K ())) . unSOP . from
```
Comment (phadej) [inline src/Generics/SOP/Universe.hs]: I to don't expect much use for these. and naming these is hard. One option is to have `Generics.SOP.Extra` module with more controversial and name-clashy stuff. Like http://hackage.haskell.org/package/lens-4.16/docs/Control-Lens-Extras.html

**Comment by kosmikus** [comment]
@phadej Thanks for the comments. Moving the extra conversions into a separate module seems a bit arbitrary to me, and not very discoverable. So you'd remove them rather than have them under their current name?

**Comment by phadej** [comment]
@kosmikus I'm neutral. I personally almost inclusively use `Generics.SOP` qualified, so for me name clash with `Prelude` names is not a big deal. (`from`, `to` also clash with `lens`, in addition to `GHC.Generics`)

**Comment by phadej** [comment]
FWIW, I have weird intuition that `IsSum` would be isomorphic to `NS I xs`. That's quite rare configuration, isn't it? (`Either`?)

**Comment by kosmikus** [comment]
@phadej I'm not sure what you're saying, sorry. Could you please elaborate?

**Comment by phadej** [comment]
About: `IsSum`, just ignore :)

about previous, `fromEnum` is consistent name, and I won't worry about clash with Prelude.

**Comment by kosmikus** [comment]
@phadej If we keep the Prelude-name-clashing export, then it'd probably be fair to make it a major version bump even if not otherwise required. Would you agree?

**Comment by phadej** [comment]
I don't like major bumps. I'd rather test (= try to compile) reverse dependencies, and make revisions to the few (if there are only few) packages which have a problem.

I can do the test early next week.

**Comment by kosmikus** [comment]
That's also a good idea.

**Comment by kosmikus** [comment]
What about resolving the name clash by using:
```
enumTypeFrom
enumTypeTo
productTypeFrom
productTypeTo
```
instead of
```
enumFrom
enumTo
productFrom
productTo
```
?

**Comment by phadej** [comment]
or `enumRepFrom`... naming is difficult.

**Comment by kosmikus** [comment]
@phadej The idea of using `Type` is that it matches the `IsEnumType` and `IsProductType` synonyms.

**Pull request merged**